Source/WebCore/html/InputType.cpp

```diff
@@ -142,27 +149,38 @@ static InputTypeFactoryMap createInputTypeFactoryMap()
     };
 
     InputTypeFactoryMap map;
-    for (auto& inputType : inputTypes)
-        map.add(inputType.nameFunction(), std::make_pair(inputType.conditionalFunction, inputType.factoryFunction));
+    for (auto& inputType : inputTypes) {
+        auto name = inputType.nameFunction();
+        map.add(name, InputTypeFactory { name, inputType.conditionalFunction, inputType.factoryFunction });
+    }
     return map;
 }
 
-static inline std::pair<InputTypeConditionalFunction, InputTypeFactoryFunction> findFactory(const AtomString& typeName)
+static inline InputTypeFactory* findFactory(const AtomString& typeName)
 {
     static NeverDestroyed factoryMap = createInputTypeFactoryMap();
-    auto factory = factoryMap.get().get(typeName);
-    if (UNLIKELY(!factory.second))
-        factory = factoryMap.get().get(typeName.convertToASCIILowercase());
-    return factory;
+    auto& map = factoryMap.get();
+    auto it = map.find(typeName);
+    if (UNLIKELY(it == map.end())) {
+        it = map.find(typeName.convertToASCIILowercase());
+        if (it == map.end())
+            return nullptr;
+    }
+    return &it->value;
 }
 
-Ref<InputType> InputType::create(HTMLInputElement& element, const AtomString& typeName)
+Ref<InputType> InputType::createIfDifferent(HTMLInputElement& element, const AtomString& typeName, InputType* currentInputType)
```

Comment:
I find it a little bit strange that this function called `createIfDifferent` can return the `currentInputType` object that was passed in, which we then assign back to the same variable we got it from. We end up bumping the refcount up and down in that case, too.

What if instead we make this function return `RefPtr<InputType>`, return `nullptr` if the type already matches, and have `HTMLInputElement::updateType` do:

```c++
auto newType = InputType::createIfDifferent(...);
if (!newType)
    return;
```

Reply:
Sure, that was an alternative design I was thinking of anyway. So done.